integration_test/warehouse/integration_test.go

```diff
@@ -0,0 +1,3462 @@
+package warehouse_test
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/warehouse/validations"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+
+	"github.com/rudderlabs/rudder-server/warehouse/integrations/tunnelling"
+
+	"github.com/rudderlabs/compose-test/compose"
+	"github.com/rudderlabs/compose-test/testcompose"
+
+	"go.uber.org/atomic"
+
+	"github.com/rudderlabs/rudder-server/services/controlplane/identity"
+
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-server/testhelper/health"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+
+	"github.com/rudderlabs/rudder-server/admin"
+
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	mocksBackendConfig "github.com/rudderlabs/rudder-server/mocks/backend-config"
+	"github.com/rudderlabs/rudder-server/utils/pubsub"
+
+	"github.com/google/uuid"
+
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	whclient "github.com/rudderlabs/rudder-server/warehouse/client"
+	whutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+
+	sqlmw "github.com/rudderlabs/rudder-server/warehouse/integrations/middleware/sqlquerywrapper"
+
+	"github.com/golang/mock/gomock"
+	"github.com/ory/dockertest/v3"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/rudderlabs/rudder-go-kit/filemanager"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats/memstats"
+	kithelper "github.com/rudderlabs/rudder-go-kit/testhelper"
+	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource"
+	"github.com/rudderlabs/rudder-server/app"
+	"github.com/rudderlabs/rudder-server/enterprise/reporting"
+	mocksApp "github.com/rudderlabs/rudder-server/mocks/app"
+	"github.com/rudderlabs/rudder-server/warehouse"
+)
+
+const (
+	succeeded    = "succeeded"
+	waiting      = "waiting"
+	aborted      = "aborted"
+	exportedData = "exported_data"
+)
+
+func TestUploads(t *testing.T) {
+	admin.Init()
+	validations.Init()
+
+	t.Run("tracks loading", func(t *testing.T) {
+		const (
+			workspaceID           = "test_workspace_id"
+			sourceID              = "test_source_id"
+			destinationID         = "test_destination_id"
+			destinationRevisionID = "test_destination_id"
+			namespace             = "test_namespace"
+		)
+
+		pool, err := dockertest.NewPool("")
+		require.NoError(t, err)
+
+		pgResource, err := resource.SetupPostgres(pool, t)
+		require.NoError(t, err)
+		minioResource, err := resource.SetupMinio(pool, t)
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		webPort, err := kithelper.GetFreePort()
+		require.NoError(t, err)
+
+		backendConfigCh := make(chan pubsub.DataEvent)
+
+		g, gCtx := errgroup.WithContext(ctx)
+		g.Go(func() error {
```

Comment:
It was basically for [this](https://github.com/rudderlabs/rudder-server/blob/4610c7ed670c54e020b33bbc970fa78caecc500c/warehouse/app.go#L443) reason since the backend config channel get's closed, it shut down the routers. I am thinking of waiting for the context here, then shut down. wdyt?


```GO
// Gets the config from config backend and extracts enabled write keys
func (a *App) monitorDestRouters(ctx context.Context) error {
	dstToWhRouter := make(map[string]*router.Router)

	ch := a.tenantManager.WatchConfig(ctx)
	for configData := range ch {
		err := a.onConfigDataEvent(ctx, configData, dstToWhRouter)
		if err != nil {
			return fmt.Errorf("config data event error: %v", err)
		}
	}

	<-ctx.Done()

	g, _ := errgroup.WithContext(context.Background())
	for _, r := range dstToWhRouter {
		r := r
		g.Go(r.Shutdown)
	}
	return g.Wait()
}
```

Reply:
Or we can keep the existing code as it is and modify in tests to close the channel once the context is cancelled.

```GO
        bcConfigWg := sync.WaitGroup{}

	mockBackendConfig := mocksBackendConfig.NewMockBackendConfig(mockCtrl)
	mockBackendConfig.EXPECT().WaitForConfig(gomock.Any()).DoAndReturn(func(ctx context.Context) error { return nil }).AnyTimes()
	mockBackendConfig.EXPECT().Identity().Return(&identity.NOOP{})
	mockBackendConfig.EXPECT().Subscribe(gomock.Any(), backendconfig.TopicBackendConfig).DoAndReturn(func(ctx context.Context, topic backendconfig.Topic) pubsub.DataChannel {
		ch := make(chan pubsub.DataEvent, 1)
		ch <- pubsub.DataEvent{
			Data:  bcConfig,
			Topic: string(backendconfig.TopicBackendConfig),
		}

		bcConfigWg.Add(1)
		go func() {
			defer bcConfigWg.Done()

			<-ctx.Done()
			close(ch)
		}()

		return ch
	}).AnyTimes()
        .
        .
        .


	bcConfigWg.Wait()
	return nil
}
```

Reply:
I'm cool with the test changes :+1: